### ISSUE TITLE
Extremely slow python iteration on CPU vs NumPy

### ISSUE THREAD

**Issue by narfanar** [pytorch/pytorch#43153]
## To Reproduce

PyTorch:
```py
import torch
from tqdm import tqdm
rows = torch.rand(2_000_000, 2)
_ = [[str(x) for x in row] for row in tqdm(rows)]
#^ Takes 7 seconds to _start_
#^ Averages 2000 it/s
```

NumPy:
```py
import numpy as np
from tqdm import tqdm
rows = np.random.rand(2_000_000, 2)
_ = [[str(x) for x in row] for row in tqdm(rows)]
#^ Starts immediately
#^ Averages 270_000 it/s. Around 130x faster.
```

## Expected behavior

Iteration (access?) isn't as horribly slow as it is now. To allow for useful Pythonic code outside training without having to convert to NumPy arrays.

## Environment
 - PyTorch Version (e.g., 1.0): pytorch-1.7.0.dev20200817
 - OS (e.g., Linux): Linux / Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): N/A
 - Python version: 3.7.7
 - CUDA/cuDNN version: 11.0 / v8.0.1.13
 - GPU models and configuration: GTX Titan X * 8
 - Any other relevant information: N/A

## Additional context

N/A

**Comment by colesbury** [comment]
Nearly all the time in the example is spent calling `str(x)`, not in the iteration over tensors.

**Comment by colesbury** [comment]
For slow iteration see https://github.com/pytorch/pytorch/issues/43116 and https://github.com/pytorch/pytorch/issues/29973. It's roughly ~8x slower than NumPy, not 100x.

**Comment by narfanar** [comment]
I edited the number typo.
`str(x)` is shared in both snippets. Can you elaborate why it would have a variable effect? @colesbury

**Comment by colesbury** [comment]
Yes, there are different `str` implementations in PyTorch and NumPy. I don't think there's much reason to optimize the `str` implementation. If `str` performance is important for your use case, please open an issue for that.

We'd like to make indexing have less overhead, but I don't think there's an easy fix. One reason is that "scalar" values (i.e. 0-dim tensors that contain only a single number) have more state in PyTorch than in NumPy, because in PyTorch they can track autograd state and are views of the underlying tensor (instead of immutable values).

Let's track the slow indexing in #29973.
